compiler/rustc_infer/src/infer/resolve.rs

```diff
@@ -39,7 +40,7 @@ impl<'a, 'tcx> TypeFolder<TyCtxt<'tcx>> for OpportunisticVarResolver<'a, 'tcx> {
             t // micro-optimize -- if there is nothing in this type that this fold affects...
         } else {
             let t = self.shallow_resolver.fold_ty(t);
-            t.super_fold_with(self)
+            ensure_sufficient_stack(|| t.super_fold_with(self))
```

Comment:
there's a very high chance that this will be perf negative 💀

Reply:
Can you start a perf run to see how bad it is?

Reply:
The new version of this code, with the `fold_ty` call before the `has_non_region_infer` condition, is very counter-intuitive, and seems to defeat half the purpose of the condition. I suggest reverting to the previous version, and then I think it's good to merge.

Reply:
Well, the idea was that `ShallowResolver::fold_ty` already checks for `t == ty::Infer` before it does anything, but I've reverted this change for now.

I still think there's some potential for (re-)optimizations here, but I'll leave that for a future PR.